controllers/appgroup_controller.go

```diff
@@ -526,35 +540,55 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
 		for _, node := range wf.Status.Nodes {
 			nodes[node.ID] = node
 		}
-		graph, err := workflow.Build(g.Name, nodes)
-		if err != nil {
-			logr.Error(err, "failed to build the wf status DAG")
+		rwf := &v1alpha12.Workflow{
+			ObjectMeta: v1.ObjectMeta{Labels: make(map[string]string)},
 		}
 
-		rev := graph.Reverse()
-
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = pkg.HelmUninstall(hr.Spec.ReleaseName, hr.Spec.TargetNamespace)
+		rwfName := fmt.Sprintf("%s-reverse", wf.Name)
+		rwfNamespace := wf.Namespace
+		err := r.Client.Get(ctx, types.NamespacedName{Namespace: rwfNamespace, Name: rwfName}, rwf)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				err = r.generateReverseWorkflow(ctx, logr, nodes, &g)
 				if err != nil {
-					logr.Error(err, "failed to uninstall helm release using helm actions - continuing with cleanup")
+					logr.Error(err, "failed to generate reverse workflow")
+					return err
 				}
-			}
-		}
 
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
+				// update the forward workflow metadata with a finalizer
+				wf.Finalizers = []string{rwffinalizer}
```

Comment:
You will need to publish the update to add the finalizer to the actual Workflow object otherwise it will not prevent the deletion of the workflow. You can use the patching logic,

```go
	wfPatch := client.MergeFrom(wf.DeepCopy())
	_ = r.Client.Patch(ctx, &wf, wfPatch)
```
and then delete the object

Reply:
Thanks.